### DISCUSSION TITLE
Ability to invite users to a private room based on an invite link.

### DISCUSSION THREAD

**Discussion by ara4n** [element-hq/element-meta#574] [Ideas]
Rather than inviting users to a private room by their email address, it could be really nice to let them join via an alias but provide a password (or a "magic link").

**Comment by dbkr** [comment]
As in a single-use link or one they could also give to others?

**Comment by KopfKrieg** [comment]
I wouldn't make it single use. Single use means I click it, and then it works exactly once. But what if I want to join the conversation later again, or I need to switch to a different browser, or I joined via phone and now want to join via laptop too?

Most conference/chat systems I know allow the link to be used multiple times (and therefore even to share it).

**Comment by MurzNN** [comment]
As example of UX implementation, we can lookup Skype, that already have invite links to private groups, also allowing joining to room as guest account (only fill name, without registering)!

**Comment by MurzNN** [comment]
Also we have [knock feature](https://github.com/matrix-org/matrix-doc/pull/2403) PR, that can be combined with invite links, so when user open invite link, he generate "knock" event, and room moderator can accept or reject it.

**Comment by hex-m** [comment]
Here is the issue for the knock feature: https://github.com/vector-im/element-web/issues/14237

**Comment by cdauth** [comment]
This seems to be the one thing that is still blocking me from migrating from Mattermost to Matrix.

My use cases are that I am part of several loose organizations that have some kind of internal communication infrastructure, such as a private Wiki or a mailing list. For the organization, I want to create a private Matrix space or room. As far as I understand, right now to invite new people to that space/room, they have to provide me with their Matrix username and I have to manually invite them to the space/room. This centralizes the responsibility and power of adding people, which goes against the principles of my organizations. Instead, I would like to generate an invite link that I can share over the internal communication infrastructure of my organizations and that permanently enables people to join the space/room after they have created a Matrix account.

There may be other use cases where it is undesirable for the link to be permanently valid, for example when using a link to invite a single person through another communication channel (such as Signal). Ideally, the UI would allow specifying the validation period of the invite link, and it would list all existing valid invite links along with the possibility to revoke them. But for a start, I believe permanent invite links are the most needed.

**Comment by q9f** [comment]
We have been using matrix (vector, riot, element) for many years now for companies internally, but also for larger public communities. It has matured a lot and clients like element are so much more usable today as compared to five years ago; and we really appreciate it.

However, I have to agree with @cdauth that this issue is potentially the biggest pain point in managing communities currently. As a space admin, I need to be able to invite users to a private room without knowing their handles. Currently, you have to ask each individual for their matrix handle which is a drastic overhead for groups of 50+.

An invite link would be the most smooth solution and I hope that this can be implemented soon. ❤️

**Comment by shamiv** [comment]
Hi all. I agree with what has been said in the previous two posts. Are there any developments in this respect?

**Reply by kurahaupo** [reply]
It occurs to me that if there's an existing invitation facility that provides "join once" links, to implement a permanent (or timed) link, you could have a bot that's also accessible as a web page, that creates a new invitation on the fly and redirects to it.

Then you just kick the bot when you want to revoke the permanent link.

**Comment by shamiv** [comment]
This is how Signal does it. It would be great to have something similar in Matrix. (Is this the right place for discussing this btw? Or is this something that should be discussed on the git van [matrix.org](https://github.com/matrix-org)?

[Screencast from 03-05-23 17:57:53.webm](https://user-images.githubusercontent.com/26513701/235976998-069ca793-6e20-4656-ab01-2c07ad083b9d.webm)

**Comment by snowinmars** [comment]
So, a user can write to a bot - and the bot could add the user to a room, like in telegram.
Are there any bots that can do this?

UPD: Seems like [matrix-bot-sdk](https://turt2live.github.io/matrix-bot-sdk/index.html) can do it.

```js
import { MatrixClient, SimpleFsStorageProvider, AutojoinRoomsMixin } from "matrix-bot-sdk";

const homeserverUrl = 'https://matrix.org';
// const client = await auth.passwordRegister("username", "password");
// console.log("Copy this access token to your bot's config: ", client.accessToken);
const accessToken = '...';
const targetRoom = '!id:matrix.org' // settings -> advanced -> internal room id
const storage = new SimpleFsStorageProvider('botStorage.json');

const client = new MatrixClient(homeserverUrl, accessToken, storage);
AutojoinRoomsMixin.setupOnClient(client);

client.on('room.event', handleRoomEvent)

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

function msToTime(ms) {
  let seconds = (ms / 1000).toFixed(1);
  let minutes = (ms / (1000 * 60)).toFixed(1);
  let hours = (ms / (1000 * 60 * 60)).toFixed(1);
  let days = (ms / (1000 * 60 * 60 * 24)).toFixed(1);
  if (seconds < 60) return seconds + " seconds";
  else if (minutes < 60) return minutes + " minutes";
  else if (hours < 24) return hours + " hours";
  else return days + " days"
}

async function handleRoomEvent(roomId, event) {
  try {
    const allowedTypes = ['m.room.encrypted', 'm.room.message'];

    if (!allowedTypes.includes(event.type)) {
      console.log('Skip uninteresting event', event.type);
      return;
    }

    const isMyEvent = event['sender'] === await client.getUserId()
    if (isMyEvent) {
      console.log('Skip my event');
      return;
    }

    await client.inviteUser(event.sender, targetRoom);
  } catch (e) {
    if (e.errcode === 'M_LIMIT_EXCEEDED') {
      await client.replyNotice(roomId, event, `Server overload, retry after ${msToTime(e.retryAfterMs)}.`);
      console.log('Too many requests, awaiting', e.retryAfterMs, 'ms');
      await sleep(e.retryAfterMs);
      console.log('...awaited');
    } else {
      console.error(e);
      await client.replyNotice(roomId, event, `Unknown error. Contact developers.`);
    }
  }
}

client.start().then(() => console.log('Bot started'));
```

**Comment by jklinkenberg** [comment]
A feature to invite users to a private space or room would be really nice.
Maybe with a link that has a configurable expiration date.
The absence of such a feature is currently preventing us from using Matrix/Element in a larger project with many partners and we will most likely need to use Mattermost where such a workflow is available

**Comment by unode** [comment]
In the space management Invite interface, there's currently a "share this space" link above the Invite search bar.
This generates an invite link but when visited by a new user, it errors informing that you aren't a member of that space/room.

Is the discussion here to allow this link to work as an invite?
Effectively making possible the use-case: "If you know this link, you can join this space/channel"

**Reply by mibressler** [reply]
Saves a lot of time & friction when onboarding

**Reply by basic-settings** [reply]
This would be extremely useful, we're looking for alternatives to Whatsapp for our 200-400 members community and an invite link is just what we need.

It's a shame this is such an old feature request that hasn't seen any movement.

**Reply by kurahaupo** [reply]
> Is the discussion here to allow this link to work as an invite? Effectively making possible the use-case: "If you know this link, you can join this space/channel"

I would hope that the solution to this would offer more fine-grained control:
* include a validity period (as timestamps for
  * `not-valid-before` and
  *  `not-valid-after`
* include control flags:
  * `once` ― tie an invitation to a sign-in identity on first use, so that:
    * multiple people cannot use it, but one person can still join from multiple devices; and therefore as a corollary:
    * the issuer can retract an invitation by tying it to a dummy (or their own) identity.
* An agent in a space/channel can, in response to someone knocking with an invitation:
  * automatically approve them if their invitation includes a token specified by the issuer;
  * record their identity and making it available to the invitation issuer.
  (presumably such an agent would be part of the server, though in principle it could be separate)

**Comment by pseudomonas** [comment]
I'm in a group of about 80 people; Want to leave discord; I think Matrix would be a good fit. I can't invite everyone by email because I don't actually know their emails and it would be a privacy issue to ask. 

This issue's been around for nearly six years so I don't think anything's gonna happen; so my main question is, has anyone here found a workaround or strategy that makes things less annoying than "create an account, then  send me your account name through another channel, then I invite you"? 

Tempted to create a second public space with a uninformative-and-hard-to-guess name, send people a link to that, and then set _that_ space to private and and say "anyone in that space automatically has access to this space".  But there's probably some way that public rooms can be listed even if the name's not guessable? Maybe if the public room is on a small server where few people are likely to be trying to enumerate the rooms?

**Reply by hex-m** [reply]
Knocking (ask to join) might be an option.

You could also make the room public for some days, distribute the link in your Discord and change it to invite-only afterwards (and remove accounts that shouldn't actually be there).

> But there's probably some way that public rooms can be listed even if the name's not guessable?

Public room usually means "everyone can join" and "listed in the room directory of the server" but the two things can be configured independently. And which room alias(es) are created for the room is a third independent thing.

No, there is generally no way to list all "anyone can join" rooms.

**Reply by Shurikn** [reply]
With the current discord exodus, having this simple feature would be really important.

**Reply by vrisalab** [reply]
There is [this project](https://github.com/dfuchss/matrix-joinlink), that creates a public room where anyone who joins is invited to a private room (could be a space too). It's not perfect but could serve your use case
